File: kafka-client-ex/src/main/java/in/srini91/learn/kafkaclientex/controller/PersonController.java
package in.srini91.learn.kafkaclientex.controller;

import in.srini91.learn.kafkaclientex.dao.model.Person;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.util.concurrent.ListenableFuture;
import org.springframework.util.concurrent.ListenableFutureCallback;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Date;

@RestController
@RequestMapping(path = "person")
public class PersonController {
    private static Logger LOG = LogManager.getLogger(PersonController.class);

    @Autowired
    private KafkaTemplate<String,Object> kf;

    @PostMapping
    public void publishPerson(@RequestBody Person person){
        ListenableFuture<SendResult<String, Object>> sendPerson = kf.send("person-topic", person.getName(), person);
        sendPerson.addCallback(new ListenableFutureCallback<SendResult<String, Object>>() {
            @Override
            public void onFailure(Throwable ex) {
                LOG.error(ex.getMessage());
            }

            @Override
            public void onSuccess(SendResult<String, Object> result) {
                LOG.info("Records Metatdata \n" +
                        "Topic : " + result.getRecordMetadata().topic() + "\n" +
                        "Partition : " + result.getRecordMetadata().partition() + "\n" +
                        "Offset : " + result.getRecordMetadata().offset() + "\n" +
                        "Timestamp : " + new Date(result.getRecordMetadata().timestamp()));
            }
        });
    }
}

File: kafka-cloud-stream-ex/src/main/kotlin/srini/learn/kafka/kafkacloudstreamex/controller/PersonController.kt
package srini.learn.kafka.kafkacloudstreamex.controller

import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.SendResult
import org.springframework.util.concurrent.ListenableFutureCallback
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import srini.learn.kafka.kafkacloudstreamex.dao.model.Person
import java.util.*

@RestController
@RequestMapping(path = ["person"])
class PersonController {

    private val LOG = LogManager.getLogger(PersonController::class.java)


//    @Autowired
//    private val kf: KafkaTemplate<String, Any>? = null
//
//    @PostMapping
//    fun publishPerson(@RequestBody person: Person) {
//        person?.name?.let { kf!!.send("person-topic", it, person) }
//            ?.addCallback(object : ListenableFutureCallback<SendResult<String?, Any?>?> {
//                override fun onFailure(ex: Throwable) {
//                    LOG.error(ex.message)
//                }
//
//                override fun onSuccess(result: SendResult<String?, Any?>?) {
//                    LOG.info(
//                        """
//                            Records Metatdata
//                            Topic : ${result!!.recordMetadata.topic()}
//                            Partition : ${result.recordMetadata.partition()}
//                            Offset : ${result.recordMetadata.offset()}
//                            Timestamp : ${Date(result.recordMetadata.timestamp())}
//                            """.trimIndent()
//                    )
//                }
//            })
//    }
}